Tools/paletteable sprite editor/js/editor.js

```diff
@@ -0,0 +1,473 @@
+var canvasWidthInPixels = 200,
+    canvasHeightInPixels = 200,
+    pixelWidth = 16,
+    pixelHeight = 16,
+    cursorOffset = 0.5,
+    cursorLineW = 2,
+    pixelRenderColor = "#808080",
+    pixelRawColor = "#002080",
+    drawPosRaw = [],
+    mark,
+    swatches = [
+"#000000",
+"#FFFFFF",
+"#FF0000",
+"#00FF00",
+"#0000FF",
+"#FFFF00",
+"#00FFFF",
+"#FF00FF"];
+
+
+Color = function (hexOrObject) {
+    var obj;
+    if (hexOrObject instanceof Object) {
+        obj = hexOrObject;
+    } else {
+        obj = LinearColorInterpolator.convertHexToRgb(hexOrObject);
+    }
+    this.r = parseInt(obj.r);
+    this.g = parseInt(obj.g);
+    this.b = parseInt(obj.b);
+};
+Color.prototype.asRgbCss = function () {
+    return "rgb(" + this.r + ", " + this.g + ", " + this.b + ")";
+};
+
+function rgbToHex(rgb) {
+    var hex = Number(rgb).toString(16);
+    if (hex.length < 2) {
+        hex = "0" + hex;
+    }
+    return hex;
+}
+
+function logicalToColor(indexA, indexB, interp, paletteSize) {
+    return new Color({
+        r: Math.min(255,Math.floor((indexA + 0.5) / paletteSize * 256.0)),
+        g: Math.min(255,Math.floor((indexB + 0.5) / paletteSize * 256.0)),
+        b: Math.floor(interp * 255.0) });
+}
+
+
+Color.prototype.asHex = function () {
+    return "#" + rgbToHex(this.r) + rgbToHex(this.g) + rgbToHex(this.b);
+};
+
+Color.prototype.toRendered = function (paletteSize) {
+    // console.log(paletteSize);
+    const result = this.toLogical(paletteSize)
+
+    const indexA = result.indexA;
+    const indexB = result.indexB;
+    const interp = result.interp;
+
+    const paletteColorA = new Color(swatches[indexA]);
+    const paletteColorB = new Color(swatches[indexB]);
+
+    //console.log("Palette color A: " + paletteColorA.asHex());
+    //console.log("Palette color B: " + paletteColorB.asHex());
+    //console.log("lerp: " + interp);
+
+    return LinearColorInterpolator.findColorBetween(
+        paletteColorA,
+        paletteColorB,
+        interp * 100.0);
+};
+
+Color.prototype.toLogical = function (paletteSize) {
+    // console.log(paletteSize);
+    var indexA = Math.floor(Math.min(this.r / 255.0, .99) * paletteSize);
```

Comment:
Codacy found an issue: [The numeric literal '.99' will have at different value at runtime.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6395467)